nomad/consul.go

```diff
@@ -92,10 +101,10 @@ func (sir ServiceIdentityRequest) Description() string {
 // ACL requirements
 // - acl:write (transitive through ACLsAPI)
 type ConsulACLsAPI interface {
-
-	// CheckSIPolicy checks that the given operator token has the equivalent ACL
-	// permissiveness that a Service Identity token policy for task would have.
-	CheckSIPolicy(ctx context.Context, task, secretID string) error
+	// CheckPermissions checks that the given Consul token has the necessary ACL
+	// permissions for each way that Consul is used as indicated by usage,
+	// returning an error if not.
+	CheckPermissions(ctx context.Context, namespace string, usage *structs.ConsulUsage, secretID string) error
```

Comment:
This left me with a belief that we need to overhaul how Nomad handles Consul tokens. In the past the Nomad agent token was used no matter who the job submitter or what the job contained. With Connect we introduced a Consul ACL token permissions check that the job submitter had equivalent Service Identity permissions, because such a token is generated and given to the task. With Namespaces, we need to check that the job submitter has necessary service:write and kv:read permissions within their namespace - and without extensive plumbing the only safe way to do that is to parse and check the HCL ourselves. Ideally we should be actually using the job submitter's Consul ACL token and letting Consul do the ACL checks - but to do that safely requires managing AccessorID's in the Nomad Client, wiring up some RPC's with Nomad Server to get the SecretID from Consul and passing it back to the Client, who can then submit the token to Consul without it ever being written to the Client's disk (the security hole @schmichael points out). That seems like a lot of work so it'll have to wait for another day.